src/providers/go.rs

```diff
@@ -72,3 +77,79 @@ impl Provider for GolangProvider {
         )])))
     }
 }
+
+impl GolangProvider {
+    pub fn read_go_mod_if_exists(&self, app: &App) -> Result<Option<String>> {
+        if app.includes_file("go.mod") {
+            Ok(Some(app.read_file("go.mod")?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn get_nix_golang_pkg(go_mod_contents: Option<String>) -> Result<String> {
+        if go_mod_contents.is_some() {
+            let mut lines = go_mod_contents.as_ref().unwrap().lines();
+            let go_version_line = lines.find(|line| line.trim().starts_with("go"));
+
+            if let Some(go_version_line) = go_version_line {
+                let go_version = go_version_line.trim().split_whitespace().nth(1).unwrap();
+
+                if let Some(nix_pkg) = version_number_to_pkg(go_version)? {
+                    return Ok(nix_pkg);
+                }
+            }
+        }
+
+        Ok(DEFAULT_GO_PKG_NAME.to_string())
+    }
+}
+
+fn version_number_to_pkg(version: &str) -> Result<Option<String>> {
+    let matched_version = AVAILABLE_GO_VERSIONS.iter().find(|(v, _)| v == &version);
+
+    match matched_version {
+        Some((_, pkg)) => Ok(Some(pkg.to_string())),
+        None => bail!("Go version {} is not available", version),
```

Comment:
I think if the version isn't specified we should default to the `go` package. Thoughts?

Reply:
I agree, this should be how we handle it for all providers. If the version isn't found, default to the version on the nixos tag (like the node provider).

Reply:
The node provider will actually error out if you specify a version in engines that's not in the list of supported versions (I tested the behavior myself). I wanted to preserve the same behavior as on node. But I can change this later.

Reply:
There's a difference between specifying unsupported version and not specifying any version. Node provider falls back to [default version](https://github.com/railwayapp/nixpacks/blob/fcf36f29ab8455f10a4c546ea6c58a9f296c197c/src/providers/node.rs#L156) if version isn't specified.